ambry-tools/src/main/java/com/github/ambry/clustermap/HelixFullAutoReconstructResourceTool.java

```diff
@@ -0,0 +1,336 @@
+        for (int k = 0; k < subCliqueInfo.length(); k++) {
+          JSONObject hostInfo = subCliqueInfo.getJSONObject(k);
+          String host = hostInfo.getString("host");
+          resourceToHosts.get(resourceId).add(host);
+          if (!hostToResources.containsKey(host)) {
+            hostToResources.put(host, new ArrayList<>());
+          }
+          hostToResources.get(host).add(resourceId);
+        }
+      }
+    }
+
+    // Validate host is present only in 1 resource such as host1 -> [resource 1], host2 -> [resource 2]..
+    for (Map.Entry<String, List<Integer>> entry : hostToResources.entrySet()) {
+      String host = entry.getKey();
+      List<Integer> resources = entry.getValue();
+      if (resources.size() != 1) {
+        throw new IllegalStateException(
+            "Input is invalid. Host " + host + "is present in more than 1 resource " + resources);
+      }
+    }
+
+    for (Map.Entry<Integer, List<String>> entry : resourceToHosts.entrySet()) {
+      Integer resourceId = entry.getKey();
+      List<String> hosts = entry.getValue();
+      System.out.println("Resource = " + resourceId + " has " + hosts.size() + " hosts. Hosts are " + hosts);
+    }
+  }
+
+  /**
+   * Build resource to partitions map
+   * Resource 1 -> [P1, P2,... P100]
+   */
+  public void buildResourceToPartitionsMap() {
+
+    // 1. Build host to partitions map
+    // Host1 -> [P1, P2... P13]
+    List<String> resourcesInCluster = admin.getResourcesInCluster(helixClusterName);
+    for (String resourceName : resourcesInCluster) {
+      IdealState idealState = admin.getResourceIdealState(helixClusterName, resourceName);
+      Map<String, List<String>> preferenceLists = idealState.getPreferenceLists();
+      this.preferenceLists.putAll(preferenceLists);
+      for (Map.Entry<String, List<String>> entry : preferenceLists.entrySet()) {
+        String partitionName = entry.getKey();
+        List<String> instances = entry.getValue();
+        for (String instanceName : instances) {
+          if (!currentStates.containsKey(instanceName)) {
+            currentStates.put(instanceName, new ArrayList<>());
+          }
+          currentStates.get(instanceName).add(partitionName);
+        }
+      }
+    }
+
+    // 2. Build resource to partitions map by going via resources -> hosts -> partitions
+    // Resource 10000 -> [P1, P2.. P100]
+    Map<String, Set<Integer>> partitionToResources = new HashMap<>();
+    for (Map.Entry<Integer, List<String>> entry : resourceToHosts.entrySet()) {
+      int resourceId = entry.getKey();
+      List<String> hosts = entry.getValue();
+      resourceToPartitions.put(resourceId, new HashSet<>());
+      Set<String> partitionSet = resourceToPartitions.get(resourceId);
+      for (String host : hosts) {
+        String instanceName = ClusterMapUtils.getInstanceName(host, port);
+        if (currentStates.containsKey(instanceName)) {
+          List<String> partitions = currentStates.get(instanceName);
```

Comment:
let remove the instance name form the currentStates map and verify that the currentStates map is empty after the for loop to make sure the current states map has as many hosts as the resourceToHosts map.